test/unit_tests/test_workflow.py

```diff
@@ -453,3 +458,145 @@ def whatever():
+        ("Input Name", Suspend({"steps": [1, 2], "__sub_step": "Input Name", "__step_group": "Multistep"})),
+        ("Multistep", Suspend({"steps": [1, 2], "__sub_step": "Input Name", "__step_group": "Multistep"})),
+    ]
+
+
+def test_step_group_with_inputform_resume():
+    """Step group with suspended sub step should resume from the sub step."""
+
+    @step("Sub step")
+    def sub_step():
+        return {"name_validated": True}
+
+    group = step_group("Multistep", begin >> step2 >> user_action >> sub_step)
+
+    class Form(FormPage):
+        subscription_id: UUID
+
+    @workflow("Workflow with step group step", target=Target.CREATE, initial_input_form=const(Form))
+    def test_wf():
+        return init >> step1 >> group >> step3 >> done
+
+    with WorkflowInstanceForTests(test_wf, "step_group_test_workflow"):
+        init_state = {"subscription_id": uuid4()}
+
+        result, process, step_log = run_workflow("step_group_test_workflow", init_state)
+
+        assert_suspended(result)
+        assert result.unwrap()["__sub_step"] == "Input Name"
+        step_log = [t for t in step_log if "__sub_step" not in t[1].unwrap()] + step_log[-1:]
+        resume_result, step_log = resume_workflow(process, step_log, {"name": "Some name"})
+
+        assert_complete(resume_result)
+
+        assert [(t[0].name, t[1].status) for t in step_log] == [
+            ("Start", StepStatus.SUCCESS),
+            ("Step 1", StepStatus.SUCCESS),
+            ("Multistep", StepStatus.SUSPEND),
+            ("Multistep", StepStatus.SUCCESS),
+            ("Sub step", StepStatus.SUCCESS),
+            ("Multistep", StepStatus.SUCCESS),
+            ("Step 3", StepStatus.SUCCESS),
+            ("Done", StepStatus.COMPLETE),
+        ]
+        assert_state(resume_result, {"steps": [1, 2, 3], "name": "Some name", "name_validated": True})
+
+
+def test_step_group_with_failure():
+    side_effect_counter = []
+
+    @step("Effectful step")
+    def effect_step():
+        side_effect_counter.append(1)
+        return {}
+
+    group = step_group("Multistep", begin >> effect_step >> step2 >> fail)
+
+    @workflow("Workflow with step group step", target=Target.CREATE, initial_input_form=const(FormPage))
+    def test_wf():
+        return init >> step1 >> group >> step3 >> done
+
+    with WorkflowInstanceForTests(test_wf, "step_group_test_workflow"):
+        init_state = {}
+
+        result, process, step_log = run_workflow("step_group_test_workflow", init_state)
+        assert_failed(result)
+        assert extract_error(result) == "Failure Message"
+        assert 1 == len(side_effect_counter), "Side effect should be called once"
+
+        step_log = [t for t in step_log if t[0] not in [effect_step, step2, fail]]
+
+        # # Retry workflow
+        persistent = list(filter(lambda p: not (p[1].isfailed() or p[1].issuspend() or p[1].iswaiting()), step_log))
```

Comment:
In general I stop reading/trusting tests when they contain code.....

Reply:
Woops, leftover debug code